WordPress/src/test/java/org/wordpress/android/ui/activitylog/list/filter/ActivityLogTypeFilterViewModelTest.kt

```diff
@@ -1,51 +1,131 @@
 import org.wordpress.android.ui.activitylog.list.filter.ActivityLogTypeFilterViewModel.UiState
-import org.wordpress.android.ui.activitylog.list.filter.ActivityLogTypeFilterViewModel.UiState.Content
+import org.wordpress.android.ui.activitylog.list.filter.DummyActivityTypesProvider.DummyActivityType
+import org.wordpress.android.ui.activitylog.list.filter.DummyActivityTypesProvider.DummyAvailableActivityTypesResponse
 
 @InternalCoroutinesApi
 class ActivityLogTypeFilterViewModelTest : BaseUnitTest() {
     private lateinit var viewModel: ActivityLogTypeFilterViewModel
+    @Mock private lateinit var dummyActivityTypesProvider: DummyActivityTypesProvider
 
     @Before
     fun setUp() {
-        viewModel = ActivityLogTypeFilterViewModel(TEST_DISPATCHER, TEST_DISPATCHER)
+        viewModel = ActivityLogTypeFilterViewModel(dummyActivityTypesProvider, TEST_DISPATCHER, TEST_DISPATCHER)
     }
 
     @Test
     fun `fullscreen loading shown, when screen initialized`() = test {
-        val uiStates = initObservers().uiStates
+        val uiStates = init().uiStates
 
-        viewModel.start()
+        viewModel.start(0L)
 
         assertThat(uiStates[0]).isInstanceOf(UiState.FullscreenLoading::class.java)
     }
 
     @Test
-    fun `section header gets added as first item in the list`() {
-        val uiStates = initObservers().uiStates
+    fun `available activity types fetched, when screen initialized`() = test {
+        init()
 
-        viewModel.start()
+        viewModel.start(0L)
 
-        assertThat((viewModel.uiState.value as Content).items[0])
+        verify(dummyActivityTypesProvider).fetchAvailableActivityTypes(anyOrNull())
+        verifyNoMoreInteractions(dummyActivityTypesProvider)
+    }
+
+    @Test
+    fun `section header gets added as first item in the list`() = test {
+        init()
+
+        viewModel.start(0L)
+
+        assertThat((viewModel.uiState.value as UiState.Content).items[0])
                 .isInstanceOf(ListItemUiState.SectionHeader::class.java)
     }
 
-    private fun initObservers(): Observers {
+    @Test
+    fun `content shown, when fetch available activity types completes successfully`() = test {
+        val uiStates = init(successResponse = true).uiStates
+
+        viewModel.start(0L)
+
+        assertThat(viewModel.uiState.value).isInstanceOf(UiState.Content::class.java)
+    }
+
+    @Test
+    fun `fullscreen error shown, when fetch available activity types completes with error`() = test {
+        init(successResponse = false)
+
+        viewModel.start(0L)
+
+        assertThat(viewModel.uiState.value).isInstanceOf(UiState.Error::class.java)
+    }
+
+    @Test
+    fun `available activity types fetched, when error retry action invoked`() = test {
+        init(successResponse = false)
+        viewModel.start(0L)
+
+        (viewModel.uiState.value as UiState.Error).retryAction.action!!.invoke()
```

Comment:
Warning (⚠️ ): Although this is test, I usually always suggest an alternative to double explanation marks (`!!`). What if just a question mark is used instead (`?`)? This way the test will not result in an exception (if something goes wrong) but instead it will result in failure. Unless, this is what the intention of the test is, to result in an exception in case something wrong. In that case please let me know and ignore this comment.

PS: The same applies for the test below.

Reply:
In my experience "!!" has valid usecases and overusing "?" sometime leads to a harder to find bugs. 

For example, if we use "?" here and the action is null, the test will result in an error saying something like "Wanted 2 times but was 1 time".  The developer will try to understand why we expected 2 calls, they'll need to check if the action's body is invoking the correct method (fetchAvailableActivityTypes), etc.

However, if the tests fails with a NPE in the "act" block of the test, they can see right away what's wrong.

Wdyt?

Reply:
Cool, cool, I was kind of thinking about that, thanks for letting me know that this was done by design!

I'll give it a though myself to weight pros and cons on the specifics of it later and will come back to you. But, since this how you want it to behave, please disregard my comment and proceed forward by ignoring my comment.